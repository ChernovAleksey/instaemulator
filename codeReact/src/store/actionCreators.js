import {
  GET_POSTS_ARRAY,
  GET_USERS_ARRAY,
  USER_IS_FOLLOWING,
  USER_IS_UNFOLLOWING,
  GIVING_LIKE,
  TAKING_LIKE,
  ADDING_COMMENT,
  IS_MODAL_OPEN,
  CHOSEN_TO_OPERATE,
} from "./actions";

export const getPostsArrayAC = () => async (dispatch) => {
  let postsarray = [];
  const { data } = await fetch(process.env.REACT_APP_API_URL_POSTS).then(
    (res) => res.json()
  );
  postsarray = data;
  dispatch({ type: GET_POSTS_ARRAY, payload: postsarray });
};

export const getUsersArrayAC = () => async (dispatch) => {
  let usersarray = [];
  const { data } = await fetch(process.env.REACT_APP_API_URL_USERS).then(
    (res) => res.json()
  );
  usersarray = data;
  dispatch({ type: GET_USERS_ARRAY, payload: usersarray });
};

export const isUserFollowingAC =
  ({ _id }) =>
  async (dispatch) => {
    fetch(process.env.REACT_APP_API_URL_USERS, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ id: _id, target: { isFollower: true } }),
    });
    dispatch({ type: USER_IS_FOLLOWING, payload: _id });
  };

export const isUserUnFollowingAC =
  ({ _id }) =>
  async (dispatch) => {
    fetch(process.env.REACT_APP_API_URL_USERS, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ id: _id, target: { isFollower: false } }),
    });
    dispatch({ type: USER_IS_UNFOLLOWING, payload: _id });
  };

export const givingLikeAC =
  ({ _id, likes }) =>
  async (dispatch) => {
    fetch(process.env.REACT_APP_API_URL_POSTS, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        id: _id,
        target: { hasBackground: true, likes: likes + 1 },
      }),
    });
    dispatch({ type: GIVING_LIKE, payload: _id });
  };

export const takingLikeAC =
  ({ _id, likes }) =>
  async (dispatch) => {
    fetch(process.env.REACT_APP_API_URL_POSTS, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        id: _id,
        target: { hasBackground: false, likes: likes - 1 },
      }),
    });
    dispatch({ type: TAKING_LIKE, payload: _id });
  };

export const addingCommentAC =
  ({ _id, array }) =>
  async (dispatch) => {
    fetch(process.env.REACT_APP_API_URL_POSTS, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ id: _id, target: { comments: array } }),
    });
    dispatch({ type: ADDING_COMMENT, payload: { _id, array } });
  };

export const isModalOpenAC = (payload) => ({ type: IS_MODAL_OPEN, payload });

export const chosenToOperateAC = (payload) => ({
  type: CHOSEN_TO_OPERATE,
  payload,
});
