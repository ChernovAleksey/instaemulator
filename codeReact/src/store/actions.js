export const GET_POSTS_ARRAY = "GET_POSTS_ARRAY";
export const GET_USERS_ARRAY = "GET_USERS_ARRAY";
export const USER_IS_FOLLOWING = "USER_IS_FOLLOWING";
export const USER_IS_UNFOLLOWING = "USER_IS_UNFOLLOWING";
export const GIVING_LIKE = "GIVING_LIKE";
export const TAKING_LIKE = "TAKING_LIKE";
export const ADDING_COMMENT = "ADDING_COMMENT";
export const IS_MODAL_OPEN = "IS_MODAL_OPEN";
export const CHOSEN_TO_OPERATE = "CHOSEN_TO_OPERATE";

