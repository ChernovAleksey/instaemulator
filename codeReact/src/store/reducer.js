import {
  GET_POSTS_ARRAY,
  GET_USERS_ARRAY,
  USER_IS_FOLLOWING,
  USER_IS_UNFOLLOWING,
  GIVING_LIKE,
  TAKING_LIKE,
  ADDING_COMMENT,
  CHOSEN_TO_OPERATE,
  IS_MODAL_OPEN,
} from "./actions";

const initialValue = {
  usersarray: [],
  postsarray: [],
  isModalOpen: false,
  chosenToOperate: {},
};
const reducer = (state = initialValue, action) => {
  switch (action.type) {

    case GET_POSTS_ARRAY: {
      return { ...state, postsarray: action.payload };
    }
      
    case GET_USERS_ARRAY: {
      return { ...state, usersarray: action.payload };
    }
      
    case USER_IS_FOLLOWING: {
      const newArray = [...state.usersarray];
      const index = newArray.findIndex((el) => el._id === action.payload);
      if (index !== -1) {
        newArray[index].isFollower = true;
      }
      return { ...state, array: newArray };
    }
      
    case USER_IS_UNFOLLOWING: {
      const newArray = [...state.usersarray];
      console.log(newArray);
      const index = newArray.findIndex((el) => el._id === action.payload);
      if (index !== -1) {
        newArray[index].isFollower = false;
      }
      return { ...state, array: newArray };
    }
      
    case GIVING_LIKE: {
      const newArray = [...state.postsarray];
      console.log(newArray);
      const index = newArray.findIndex((el) => el._id === action.payload);
      if (index !== -1) {
        newArray[index].hasBackground = true;
        newArray[index].likes = newArray[index].likes + 1;
      }
      return { ...state, array: newArray };
    }
      
    case TAKING_LIKE: {
      const newArray = [...state.postsarray];
      console.log(newArray);
      const index = newArray.findIndex((el) => el._id === action.payload);
      if (index !== -1) {
        newArray[index].hasBackground = false;
        newArray[index].likes = newArray[index].likes - 1;
      }
      return { ...state, array: newArray };
    }
      
    case ADDING_COMMENT: {
      const newArray = [...state.postsarray];
      const index = newArray.findIndex((el) => el._id === action.payload._id);
      if (index !== -1) {
        newArray[index].comments = action.payload.array;
      }
      return { ...state, array: newArray };
    }
      
    case IS_MODAL_OPEN: {
      return { ...state, isModalOpen: action.payload };
    }
      
    case CHOSEN_TO_OPERATE: {
      return { ...state, chosenToOperate: action.payload };
    }

    default: {
      return state;
    }
  }
};

export default reducer;
