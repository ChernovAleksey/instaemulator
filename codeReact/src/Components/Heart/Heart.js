import { givingLikeAC, takingLikeAC } from "../../store/actionCreators";
import { useDispatch} from "react-redux";
import { useState } from "react";


const Heart = ({ likes, hasBackground, _id, url }) => {
    const dispatch = useDispatch() 
    const [def, setDef] = useState(likes);
    const [background, setBackground] = useState(hasBackground)
    return (
      <div>
        <div className="post_card_image_container">
          <img
            onDoubleClick={() => {
              if (!background) {
                dispatch(givingLikeAC({ _id, likes }));
                setBackground((prev) => !prev);
                setDef((def) => def + 1);
              }
            }}
            className="post_card_image"
            src={`${url}`}
            alt=""
          />
        </div>
        {background ? (
          <div
            className="post_card_heart_marked"
            onClick={() => {
              dispatch(takingLikeAC({ _id, likes }));
              setBackground((prev) => !prev);
              setDef((def) => def - 1);
            }}
          >
            &hearts;
          </div>
        ) : (
          <div
            className="post_card_heart"
            onClick={() => {
              dispatch(givingLikeAC({ _id, likes }));
              setBackground((prev) => !prev);
              setDef((def) => def + 1);
            }}
          >
            &#9825;
          </div>
        )}
        <p className="post_card_likes">{`${def}`} Likes </p>
      </div>
    );
}
export default Heart