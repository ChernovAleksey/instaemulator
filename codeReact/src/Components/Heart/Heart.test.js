import Heart from './Heart';
import { render} from "@testing-library/react";
import { Provider} from "react-redux";
import store from "../../store/index";

const Component = () => {
  return (
    <>
      <Heart />
    </>
  );
};
const RenderComponent = () => {
  return (
    <Provider store={store}>
      <Component />
    </Provider>
  );
};
describe("Heart renders", () => {
  test("should Heart match snapshot", () => {
    const { asFragment } = render(<RenderComponent />);
    expect(asFragment()).toMatchSnapshot();
  });
});