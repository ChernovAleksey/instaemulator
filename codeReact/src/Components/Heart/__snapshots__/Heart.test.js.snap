// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heart renders should Heart match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="post_card_image_container"
    >
      <img
        alt=""
        class="post_card_image"
        src="undefined"
      />
    </div>
    <div
      class="post_card_heart"
    >
      â™¡
    </div>
    <p
      class="post_card_likes"
    >
      undefined Likes 
    </p>
  </div>
</DocumentFragment>
`;
